### PR TITLE
feat(title-filter): a `stem:` prefix, for the half a plain keyword gives away

### PR DESCRIPTION
**Pull request by artemtrofymenko**
Taking you up on #3103. Additive, default untouched, and it turned out to be worth more than migration plumbing.

## Three settings, not two

The framing that made this land differently from what either of us described: a plain keyword is loose on **both** sides, and usually only one of them was meant. `word:` and a bare keyword are not one setting seen from two directions — there is a third between them, and it is the one every example in this thread actually wanted.

```
agent         Agentforce yes   Agents yes   Reagents yes     both sides open (today)
stem:agent    Agentforce yes   Agents yes   Reagents NO      must START a word
word:agent    Agentforce NO    Agents NO    Reagents NO      the whole word only
```

## It is not a no-op today

I expected a stem marker to do nothing until the default flips, and checked before writing the docs paragraph. It already narrows:

```
stem:rust engineer   vs "Sr. Zero Trust Engineer III (6794)"   ->  NO MATCH
stem:rust engineer   vs "Rust Engineering Lead"                ->  match
```

No existing setting does that: the plain form keeps both, `word:` drops both. So this is usable against the motivating case of #3103 immediately, without touching the default — which is the "usable immediately" you named, though not by the route your sentence took.

On @IaroslavMazur's corpus, reading every positive as a stem changes 4 verdicts of 675, all four false positives (`NASTAR` ×2 for `astar`, `Zero Trust Engineer` ×2 for `rust engineer`) and nothing else.

## What it does not cover, said out loud

A keyword that **ends** a longer word rather than starting it. `automation` inside `Gebäudeautomation` needs the plain form and `stem:` will not reach it:

```
                                   plain   stem:   word:
crypto      -> Cryptocurrency       yes     yes     no
fellows     -> Fellowship           yes     yes     no
automation  -> Gebäudeautomation    yes     NO      no
```

Three of the four loose cases raised in this thread are prefix-shaped; that one is not. I did not add a fourth prefix for it: the plain default already serves it, nobody has measured a config that the pair cannot express, and a suffix marker is easier to add later than to remove. The template says this rather than leaving it to be discovered.

## Tests

`tests/title-stem-prefix.test.mjs`, 8 assertions, and they are **triples rather than pairs**. "`stem:` matches X" is satisfied by a plain substring, which matches everything `stem:` does — the property worth pinning is that it sits strictly between its neighbours. Seen red in both directions, each mutation printing proof it applied:

| reverted to | fails on |
|---|---|
| plain substring | Reagents, Encrypto, Zero Trust Engineer |
| both-side anchor (`word:` under another name) | Agentforce, Agents, Cryptocurrency, Fellowship, Rust Engineering |

Your six negative-side rows are pinned as their own block, with a comment saying why they are hand-written rather than corpus-derived: none of the 675 titles contains `Internal`, `International` or `Internship`, so a corpus diff reports zero for exactly the strings the template names as the reason the default is a substring.

And the shipped template's own verdicts are asserted unchanged, since no entry uses the prefix yet.

## The doc paragraph, and what it settled

You said the spelling is whatever survives writing the docs for it. Writing them is what produced the three-way table above and killed the alternative: `agent*` reads as a glob, which invites `*agent`, and a suffix side nothing has measured a need for. `stem:` mirrors `word:`, parses on the same path, and the paragraph explains itself in six lines.

The `word:`-becomes-a-no-op finding is in that paragraph as you asked — under an anchored default it keeps meaning only for keywords with a non-word edge, which is the `.net` case @IaroslavMazur pinned down.

## On the flip

Nothing here depends on it, and I have not assumed it happens. If it does, this is the prerequisite; if it does not, `stem:` still answers `Zero Trust Engineer` on its own. The `DATA_CONTRACT.md` point you raised — that flipping silently changes behaviour for every existing `portals.yml` and nobody gets a signal — is the part I would not want to see decided by a merge either.

`--only title-stem` 8, `--only title-filter` 30, `--only scan` 95, `--only aggregator` 11, `--only keyword` 20, SYSTEM_PATHS over 1,178 tracked files.

## Summary by CodeRabbit

- **New Features**
  - Added `stem:` title-filter keywords to match words beginning with a specified prefix while allowing longer endings.
  - Added Unicode-aware matching to avoid matching prefixes within unrelated words.
  - Empty `stem:` filters match nothing.

- **Documentation**
  - Documented the differences between `stem:`, whole-word, and plain keyword matching.

- **Tests**
  - Added coverage for boundary cases, compound words, escaping, grouped filters, negative filters, and existing template behavior.

### CODE REVIEW

File: tests/title-stem-prefix.test.mjs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+const TRIPLES = [
+  // keyword, title, plain, stem, word
+  ['agent', 'Agentforce Developer', true, true, false],
+  ['agent', 'AI Agents Manager', true, true, false],
+  ['agent', 'Account Manager Instrumentation & Reagents', true, false, false],
+  ['crypto', 'Cryptocurrency Analyst', true, true, false],
+  ['crypto', 'Encrypto Systems Lead', true, false, false],
+  ['fellows', 'Fellowship Programme Lead', true, true, false],
+  ['rust engineer', 'Rust Engineering Lead', true, true, false],
+  ['rust engineer', 'Sr. Zero Trust Engineer III (6794)', true, false, false],
```
[the remark this answers is not in the document]
Comment (artemtrofymenko) [reply tests/title-stem-prefix.test.mjs:35]: Checked both halves of this before answering, and I am declining it.

**`fellows` is not fabricated — it is this module's own canonical example.** `title-keywords.mjs:17-18` on `main`, in the comment that explains why anchoring is opt-in at all:

> `"crypto" is meant to catch "Cryptocurrency" and "fellows" to catch "Fellowship", and anchoring the whole list would silently stop both.`

That sentence is the reason `word:` is a prefix rather than the default. Using the same pair to demonstrate the setting that sits between them is continuity, not invention — and swapping it for something else would leave the module explaining `stem:` with one example and `word:` with another for the identical mechanism.

**The rule cited governs a different surface.** `AGENTS.md:40` and `modes/_shared.md:33`:

> *Keywords get reformulated, never fabricated.* … If a claim isn't backed by an in-scope file, ask the user; if they don't add it, the output goes without it.

That is about generated **application documents** — a CV or cover letter must not assert experience the user's own files do not support, which is why the finding reaches for `cv.md` and `article-digest.md`. A keyword in a title-matcher fixture is not a claim about anyone; it is an input string chosen to exercise a boundary. `Reagents`, `Encrypto` and `Gebäudeautomation` are in the same test for the same reason and are equally unbacked by `cv.md`, which is the tell that the rule is being read one layer out from where it applies.

No change. If the guideline is meant to extend to test fixtures, that is worth saying explicitly in `AGENTS.md`, because as written it would also forbid the negative examples every matcher test needs.

**Review by Scott-Emberson — APPROVED** [review]
Reviewed as code owner of `tests/`. Scope is `tests/title-stem-prefix.test.mjs`; `title-keywords.mjs` is outside my area and I read it for context.

The sentinel is two-sided, which is what makes a stemming feature safe to add. Each case pins what `stem:` must match and what it must not: `agent` does not reach `Reagents`, `crypto` does not reach `Encrypto Systems Lead`, `rust engineer` does not reach `Zero Trust Engineer`. Those trace to the left-boundary-only `stemPattern` against the both-boundary anchored pattern, and the `loosenedByStem && tightenedByStem` assertion proves `stem:` collapses into neither `word:` nor the default. The bare-keyword path is exercised in every case, and the shipped `portals.example.yml` is loaded to prove no verdict drift for entries that do not use the prefix. Over-match is guarded both directions: a bare `stem:` yields false, metacharacters are escaped, and the AND-group boundary is retained, with the compound-suffix non-coverage documented honestly.

Hermetic, reads only the template. Green on CI. Approving from my side.
